File: Top-Interview-150/Sliding-Window/longest_substring_without_repeating_characters.py
# Problem: 3. Longest Substring Without Repeating Characters
# Technique: Sliding Window
# Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/description/?envType=study-plan-v2&envId=top-interview-150


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left =  0
        windows_substring = set()
        max_len = 0

        for right in range(len(s)):
            while s[right] in windows_substring:
                windows_substring.remove(s[left])
                left  += 1

            windows_substring.add(s[right])
            max_len = max(max_len, right - left + 1)
        
        return max_len

File: Top-Interview-150/Sliding-Window/longest_substring_without_repeating_characters.rs
// Problem: 3. Longest Substring Without Repeating Characters
// Technique: Sliding Window
// Link: https://leetcode.com/problems/longest-substring-without-repeating-characters/description/?envType=study-plan-v2&envId=top-interview-150
//

use std::collections::HashSet;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let chars: Vec<char> = s.chars().collect();
        let mut left = 0;
        let mut max_len = 0;
        let mut windows_substring = HashSet::new();

        for (right, &ch) in chars.iter().enumerate() {
            while windows_substring.contains(&ch) {
                windows_substring.remove(&chars[left]);
                left += 1;
            }

            windows_substring.insert(ch);
            max_len = max_len.max(right - left + 1);
        }

        max_len as i32
    }
}
